### DISCUSSION TITLE
Assessment of the difficulty in porting CPU architecture for MRPT

### DISCUSSION THREAD

**Discussion by wangyuliu** [MRPT/mrpt#1288] [Q&A] [answered]
Hello everyone! I am working on implementing a tool to assess the complexity of CPU architecture porting. It primarily focuses on RISC-V architecture porting. In fact, the tool may have an average estimate of various architecture porting efforts.My focus is on the overall workload and difficulty of transplantation in the past and future,even if a project has already been ported.As part of my dataset, I have collected the **MRPT** project. **I would like to gather community opinions to support my assessment. I appreciate your help and response!** Based on scanning tools, the porting complexity is determined to be moderate, with a moderate amount of code related to the CPU architecture in the project.  Is this assessment accurate?Do you have any opinions on personnel allocation and consumption time？ I look forward to your help and response.

**Comment by jolting** [comment, marked as answer]
MRPT is not suitable for compiling for a microcontroller that isn't running Linux or another supported operating system. Porting will be quite difficult.

Theoretically it should compile for Linux on RISC-V. However, I'm not sure it has ever been attempted, so you may run into some problems.

Linux support on RISC-V is still relatively new. You will find it much easier to use ARM. The major benefit of RISC-V is it's an open ISA, but in general ARM will outperform RISC-V solutions. ARM is also more mature, so fewer bugs. There are only a handful of RISC-V platforms that can run Linux. I would highly recommend picking an ARM capable of running Linux for an embedded application.

**Comment by jlblancoc** [comment]
Thanks @jolting ! 

Just want to add: MRPT sources, "as is", current *do build* indeed on RISC-V (and many other archs) without any further effort needed to port it:
https://buildd.debian.org/status/package.php?p=mrpt

It was possible, not without headaches over the years to fix all those little assumptions that are "forgiven" if using the "regular" archs ;-)

Anyhow, I agree with Hunter in ARM64 looking like a more reasonable architecture for embedded apps, as of today.

**Reply by jolting** [reply]
I didn't realize Debian was doing RISC-V builds of all the packages.
For ~$30 US including shipping, it might be worth getting a Mango Pi and doing some benchmarks.
https://mangopi.org/
https://bret.dk/raspberry-pi-zero-vs-mangopi-mq-pro-benchmarks/

Most comparisons of the D1 processor are to the pi-zero processor called the BCM2835. This tech is from about 2011 and is quite antiquated by today's standards. Also a PI-zero is about 40% cheaper, so it might be a better value. I think a RasPi 5 with a BCM2711 would smoke it, but I think those are going for ~$60 if you can find them.
